packages/autoupdate/autoupdate.go

```diff
@@ -0,0 +1,216 @@
+package autoupdate
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"runtime"
+	"time"
+
+	"github.com/AplaProject/go-apla/packages/conf"
+	"github.com/AplaProject/go-apla/packages/consts"
+	"github.com/AplaProject/go-apla/packages/model"
+	"github.com/AplaProject/go-apla/tools/update_client/client"
+	"github.com/AplaProject/go-apla/tools/update_client/params"
+	updateModel "github.com/AplaProject/go-apla/tools/update_server/model"
+
+	version "github.com/hashicorp/go-version"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// updates not tied to block number
+	noBlockNumber = 0
+)
+
+var updater *Updater
+
+type updateVersion struct {
+	version     string
+	blockNumber uint64
+}
+
+// Updater is updater
+type Updater struct {
+	server     string
+	pubkeyPath string
+
+	client   *client.UpdateClient
+	versions []*updateVersion
+}
+
+func (u *Updater) tryUpdate(currentBlockNumber uint64) error {
+	for _, v := range u.versions {
+		if v.blockNumber == noBlockNumber || currentBlockNumber+1 <= v.blockNumber {
+			err := model.SetStopNow()
+			if err != nil {
+				log.WithFields(log.Fields{"type": consts.DBError, "err": err}).Error("create stop daemon record")
+			}
+
+			executablePath, err := os.Executable()
+			if err != nil {
+				log.WithFields(log.Fields{"type": consts.AutoupdateError, "err": err}).Error("get application executable path")
+				return err
+			}
+
+			build, err := u.client.GetBinary(
+				params.ServerParams{Server: u.server},
+				params.KeyParams{PublicKeyPath: u.pubkeyPath},
+				params.BinaryParams{Version: v.version},
+			)
+			if err != nil {
+				log.WithFields(log.Fields{"type": consts.AutoupdateError, "err": err}).Error("get binary data")
+				return err
+			}
+
+			err = ioutil.WriteFile(executablePath, build.Body, 0755)
+			if err != nil {
+				log.WithFields(log.Fields{"type": consts.AutoupdateError, "err": err}).Error("update executable file")
+				return err
+			}
+
+			err = u.restart(executablePath)
+			if err != nil {
+				log.WithFields(log.Fields{"type": consts.AutoupdateError, "err": err}).Error("restart")
+				return err
+			}
+
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (u *Updater) restart(executablePath string) error {
+	cmd := exec.Command(executablePath)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	err := cmd.Start()
+	if err != nil {
+		return err
```

Comment:
Logging cmdError?

Reply:
This is a local method that is logged in line 74